pkg/virt-launcher/virtwrap/manager.go

```diff
@@ -1339,6 +1339,10 @@ func (l *LibvirtDomainManager) SyncVMI(vmi *v1.VirtualMachineInstance, useEmulat
 	// To make sure, that we set the right qemu wrapper arguments,
 	// we update the domain XML whenever a VirtualMachineInstance was already defined but not running
 	if !newDomain && cli.IsDown(domState) {
+		err = network.SetupPodNetworkPhase2(vmi, domain)
```

Comment:
so, we definitely call this before defining the domain for the first time. It's unclear how the domain could ever be persisted in libvirt without having this setup performed.

Are we saying the phase2 setup isn't persistent for some reason? It's unclear how this could occur.

Reply:
`phase2` setup is persisted.
The problem is that if there is a libvirt error after the domain is persisted (you can see an example of such an error in the bug [here](https://bugzilla.redhat.com/1904132)), we are trying to re-update the libvirt domain xml.
The libvirt domain xml is updated with a new version of the domain (the persisted version is not re-used).
The new version of the domain is created by invoking the `converter.Convert_v1_VirtualMachine_To_api_Domain` only.
`phase2` which in the first time invoked by `preStartHook`, is skipped in the following tries.
That's what @qinqon solves here is this PR. Calling `phase2` if the libvirt domain xml is updated (exists but not running).

Reply:
@AlonaKaplan @qinqon do we know why didn't the domain start in the first try?

Reply:
[yes](https://bugzilla.redhat.com/show_bug.cgi?id=1904132#c7).

Reply:
> The libvirt domain xml is updated with a new version of the domain (the persisted version is not re-used).

i'm surprised we have this logic at all to re-try the domain conversion if it fails the first time. this might be an old code path that shouldn't exist anymore. 

> The new version of the domain is created by invoking the converter.Convert_v1_VirtualMachine_To_api_Domain only.
> phase2 which in the first time invoked by preStartHook, is skipped in the following tries.

It's unclear to me how recalculating the domain xml and posting it again somehow resolves anything.  If the domain xml we calculated the first time failed, why would it pass after doing the same thing again later on?

I guess what I'm trying to ask is, what's the difference between the first time the domain is defined and the subsequent time it is defined due to not being able to start

Reply:
> It's unclear to me how recalculating the domain xml and posting it again somehow resolves anything. If the domain xml we calculated the first time failed, why would it pass after doing the same thing again later on?
> 
> I guess what I'm trying to ask is, what's the difference between the first time the domain is defined and the subsequent time it is defined due to not being able to start

+1 Exactly. I was trying to wrap my head around the proposed solution too and just don't understand how does it solve this issue.
Also, in comment 8, @maiqueb says that it's not possible to reproduce this.

Reply:
> i'm surprised we have this logic at all to re-try the domain conversion if it fails the first time. this might be an old code path > that shouldn't exist anymore.

Maybe the re-try logic has to be removed.
I don't know why this logic was there in the first place, and cannot think of a real life scenario the re-calculation of the domain can solve.
This PR didn't introduce the logic, it assumes the logic is there for some reason and solves a bug it has.
The re-try logic was introduced in [this](https://github.com/kubevirt/kubevirt/pull/337/commits/a22cfeeb3508b61a21ab5205f3f01a3c16b5d143) PR. @davidvossel do you remember why it was needed in the first place and what was changed so now it is not needed anymore?

> Also, in comment 8, @maiqueb says that it's not possible to reproduce this.

The original bug that discovered the issue currently cannot be reproduced since we returned the CAP_NET_ADMIN to virt-launcher.
But we are able to reproduce it using a hook sidecar that doesn't send the tap device name in the first try but sends it in the second try. (Of course that's not a real life scenario, just a way to test the fix).

Reply:
BTW, this fix won't solve the [original bug](https://bugzilla.redhat.com/190413).
The bug was caused because the hook side car has an old version and therefore `managed=false` was removed from the domain.
Even if this PR was merged, the re-tries would fail.
But while working on the bug we noticed the retries don't have the `phase2` network info on the sent domain.

Reply:
if this doesn't solve the original issue, I'm inclined to remove the retry entirely. I don't like the idea of having multiple paths for defining the domain. This code has been around since 2017. I'm betting the reason this retry exists doesn't even make sense today

@rmohr do you recall why this rety logic exists, or do you have any concerns about removing it? The context here is that it's possible for the libvirt domain to be posted, then never start, so we keep trying to post the converted domain xml in subsequent syncs

Reply:
/cc @EdDev please update the comment with the scenario you found out.

Reply:
The following scenario has been detected:
- First attempt to create the domain failed.
  With:
  ```
  {"component":"virt-launcher","kind":"","level":"error","msg":"Failed to sync vmi","name":"win-2019-standard-test-cm-9","namespace":"vm-images","pos":"server.go:161","reason":"virError(Code=1, Domain=10, Message='internal error: qemu unexpectedly closed the monitor: 2021-03-31T17:13:48.908789Z qemu-kvm: -device virtio-blk-pci-non-transitional,bus=pci.5,addr=0x0,drive=libvirt-1-format,id=ua-rootdisk-install,bootindex=2,write-cache=on: Cannot get 'write' permission without 'resize': Image size is not a multiple of request alignment')","timestamp":"2021-03-31T17:13:49.111758Z","uid":"12263510-36c9-4a6c-822f-63bfe974d255"}
  ```
 - Second attempt succeeded to create the domain, but the tap device has not been updated into the domain and therefore it allowed the domain to come up with a vnic that is not connected to a bridge.
 
Ref: https://bugzilla.redhat.com/1944418

We need either this fixed as proposed in this PR or to remove the entire operation of updating the domain spec.

I am unable to interpret what was mentioned in the original commit that introduced this special domain update ([here](https://github.com/kubevirt/kubevirt/pull/337/commits/a22cfeeb3508b61a21ab5205f3f01a3c16b5d143)), but only someone that does can actually confirm if this can be taken out or not.

Reply:
> We need either this fixed as proposed in this PR or to remove the entire operation of updating the domain spec.

I am favoring the "drop" option.
@qinqon has proposed it in #5400 with a convincing (IMO) reason.

Reply:
yes, i'm in favor of dropping this if statement as well. i acked #5200

Reply:
I am not sure if removing the retry break things more than they are, but how come that we don't get the same domain spec on retries? That sounds wrong.